pkgs/dart_tooling_mcp_server/lib/src/mixins/dtd.dart

```diff
@@ -201,6 +211,89 @@ base mixin DartToolingDaemonSupport on ToolsSupport {
+  // session.
+  Future<CallToolResult> runtimeErrors(CallToolRequest request) async {
+    return _callOnVmService(
+      callback: (vmService) async {
+        final errors = <String>[];
+        try {
+          await vmService.streamListen(EventStreams.kExtension);
+          await vmService.streamListen(EventStreams.kStderr);
+
+          // We need to listen to streams with history so that we can get errors
+          // that occurred before this tool call.
+          // TODO(https://github.com/dart-lang/ai/issues/57): this can result in
+          // duplicate errors that we need to de-duplicate somehow.
+          final extensionEvents = vmService.onExtensionEventWithHistory.listen((
+            Event e,
+          ) {
+            if (e.extensionKind == 'Flutter.Error') {
+              // TODO(https://github.com/dart-lang/ai/issues/57): consider
+              // pruning this content down to only what is useful for the LLM to
+              // understand the error and its source.
+              errors.add(e.json.toString());
+            }
+          });
+          final stderrEvents = vmService.onStderrEventWithHistory.listen((
+            Event e,
+          ) {
+            final message = decodeBase64(e.bytes!);
+            // TODO(https://github.com/dart-lang/ai/issues/57): consider
+            // pruning this content down to only what is useful for the LLM to
+            // understand the error and its source.
+            errors.add(message);
+          });
+
+          // Await a short delay to allow the error events to come over the open
+          // Streams.
+          await Future<void>.delayed(const Duration(seconds: 1));
+
+          await vmService.streamCancel(EventStreams.kExtension);
+          await vmService.streamCancel(EventStreams.kStderr);
+          await extensionEvents.cancel();
+          await stderrEvents.cancel();
+
+          if (errors.isEmpty) {
+            return CallToolResult(
+              content: [TextContent(text: 'No runtime errors found.')],
+            );
+          }
+          return CallToolResult(
+            content: [
+              TextContent(
+                text:
+                    'Found ${errors.length} '
+                    'error${errors.length == 1 ? '' : 's'}:\n',
+              ),
+              ...errors.map((e) => TextContent(text: e.toString())),
+            ],
+          );
+        } on RPCError catch (e) {
+          return CallToolResult(
+            isError: true,
+            content: [
+              TextContent(
+                text:
+                    'Failed to get runtime errors: ${e.message} '
+                    '(Code: ${e.code})',
+              ),
+            ],
+          );
+        } catch (e) {
+          return CallToolResult(
+            isError: true,
+            content: [TextContent(text: 'An unexpected error occurred: $e')],
+          );
+        }
```

Comment:
nit: add a finally clause to cancel the stream listeners (if assigned)

Reply:
moved the streamCancel and the stream subscription cancel calls to the finally block.